File: sdk/ios/Sources/SecureGatewaySDK/Session.swift
import Foundation

/// One connection session's directional keys, mirroring the Go `e2ee.Session` and the JVM
/// SDK: mobile seals with K_m2d / opens with K_d2m; desktop is the reverse. Salt is always
/// mobile-nonce-first, fixed by role.
///
/// This is a `final class` (not a `struct`) because `open` maintains per-session anti-replay
/// state (SG-02) that must persist across calls on the same instance.
public final class Session {
    private let sendKey: Data
    private let recvKey: Data

    /// Envelope ts is unix-milliseconds (the relay protocol's ts unit); an inbound envelope
    /// older than this many ms behind the highest seen ts is rejected. Mirrors the Go
    /// reference `defaultReplayWindowMillis` (5 minutes).
    private static let replayWindowMillis: Int64 = 5 * 60 * 1000

    // Anti-replay state, advanced only with id/ts the AEAD has authenticated.
    private var seen: [String: Int64] = [:]
    private var lastTs: Int64 = 0
    private var primed = false

    private init(sendKey: Data, recvKey: Data) {
        self.sendKey = sendKey
        self.recvKey = recvKey
    }

    /// Derive the directional session keys for one connection. `idPriv`/`peerIdPub` are this
    /// device's long-term identity private key and the peer's identity public key (exchanged
    /// at pairing); `ephPriv`/`peerEphPub` are this session's ephemeral private key and the
    /// peer's ephemeral public key (exchanged in the handshake). Mixing the ephemeral DH into
    /// the keying material gives forward secrecy (FR-5.2); the identity DH authenticates the
    /// peer. Mirrors the Go reference `e2ee.NewSession`.
    public static func create(idPriv: Data, peerIdPub: Data, ephPriv: Data, peerEphPub: Data,
                              role: Role) throws -> Session {
        let myEphPub = try Crypto.publicFromPrivate(ephPriv)
        // Four X25519 shared secrets (Noise-KK style):
        let ss = try Crypto.sharedSecret(myPriv: idPriv, peerPub: peerIdPub)   // identity<->identity (auth)
        let ee = try Crypto.sharedSecret(myPriv: ephPriv, peerPub: peerEphPub) // ephemeral<->ephemeral (FS)
        let md: Data
        let dm: Data
        if role == .mobile {
            md = try Crypto.sharedSecret(myPriv: idPriv, peerPub: peerEphPub)  // mobileIdentity <-> desktopEphemeral
            dm = try Crypto.sharedSecret(myPriv: ephPriv, peerPub: peerIdPub)  // desktopIdentity <-> mobileEphemeral
        } else {
            md = try Crypto.sharedSecret(myPriv: ephPriv, peerPub: peerIdPub)  // == DH(mobileIdentity, desktopEphemeral)
            dm = try Crypto.sharedSecret(myPriv: idPriv, peerPub: peerEphPub)  // == DH(desktopIdentity, mobileEphemeral)
        }
        let ikm = ss + ee + md + dm

        let mobileEphPub = role == .mobile ? myEphPub : peerEphPub
        let desktopEphPub = role == .mobile ? peerEphPub : myEphPub
        let salt = mobileEphPub + desktopEphPub

        let kM2D = Crypto.deriveKey(ikm: ikm, salt: salt, dir: "m2d")
        let kD2M = Crypto.deriveKey(ikm: ikm, salt: salt, dir: "d2m")
        return role == .mobile ? Session(sendKey: kM2D, recvKey: kD2M)
                               : Session(sendKey: kD2M, recvKey: kM2D)
    }

    /// Seal plaintext: returns nonce(24) || ciphertext+tag, ready as the envelope payload.
    public func seal(id: String, ts: Int64, plaintext: Data) throws -> Data {
        try sealWith(nonce: Crypto.randomNonce(), id: id, ts: ts, plaintext: plaintext)
    }

    /// Seal with an explicit nonce — for the deterministic interop vectors/tests.
    func sealWith(nonce: Data, id: String, ts: Int64, plaintext: Data) throws -> Data {
        let ct = try Crypto.aeadSeal(key: sendKey, nonce: nonce, aad: Crypto.aad(id: id, ts: ts), plaintext: plaintext)
        return nonce + ct
    }

    public func open(id: String, ts: Int64, wire: Data) throws -> Data {
        guard wire.count >= Crypto.nonceSize else { throw CryptoError.badLength }
        let nonce = wire.prefix(Crypto.nonceSize)
        let ct = wire.suffix(from: wire.startIndex + Crypto.nonceSize)
        let pt = try Crypto.aeadOpen(key: recvKey, nonce: Data(nonce), aad: Crypto.aad(id: id, ts: ts), cipher: Data(ct))
        // Reject duplicate delivery only after the AEAD has authenticated id/ts, so forged
        // metadata cannot advance the replay window (SG-02, FR-5.1).
        try checkReplay(id: id, ts: ts)
        return pt
    }

    private func checkReplay(id: String, ts: Int64) throws {
        if primed && ts < lastTs - Session.replayWindowMillis {
            throw CryptoError.stale
        }
        if seen[id] != nil {
            throw CryptoError.replay
        }
        seen[id] = ts
        if !primed || ts > lastTs {
            lastTs = ts
        }
        primed = true
        let floor = lastTs - Session.replayWindowMillis
        seen = seen.filter { $0.value >= floor }
    }
}

File: sdk/core/src/main/java/com/securegateway/core/Session.java
package com.securegateway.core;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

/**
 * Holds the two directional session keys for one connection session, built after
 * both handshake nonces have been exchanged. Mirrors the Go reference
 * {@code e2ee.Session}: keys are derived with HKDF-SHA256 over the X25519 shared
 * secret, with {@code salt = mobileNonce || desktopNonce} (mobile first, fixed by
 * role) and {@code info = "secure-gateway/e2ee/v1|" + dir}. Mobile seals with
 * K_m2d and opens with K_d2m; desktop is the reverse.
 */
public final class Session {

    /**
     * Envelope ts is unix-milliseconds (the relay protocol's ts unit); an inbound
     * envelope older than this many ms behind the highest seen ts is rejected.
     * Mirrors the Go reference {@code defaultReplayWindowMillis} (5 minutes).
     */
    static final long DEFAULT_REPLAY_WINDOW_MILLIS = 5L * 60L * 1000L;

    /** Thrown by {@link #open} when an envelope is a replay or outside the window (SG-02). */
    public static final class ReplayException extends RuntimeException {
        ReplayException(String message) {
            super(message);
        }
    }

    private final byte[] sendKey; // seals outbound (this device -> peer)
    private final byte[] recvKey; // opens inbound (peer -> this device)

    // Anti-replay state for inbound envelopes (SG-02), advanced only with id/ts
    // the AEAD has authenticated.
    private final long replayWindowMillis = DEFAULT_REPLAY_WINDOW_MILLIS;
    private final Map<String, Long> seen = new HashMap<>();
    private long lastTs;
    private boolean primed;

    private Session(byte[] sendKey, byte[] recvKey) {
        this.sendKey = sendKey;
        this.recvKey = recvKey;
    }

    /**
     * Derive the directional session keys. {@code mobileNonce}/{@code desktopNonce}
     * must each be {@link Crypto#HANDSHAKE_NONCE_SIZE} bytes and identical on both
     * devices.
     */
    public static Session create(byte[] myPriv, byte[] peerPub, Role role,
                                 byte[] mobileNonce, byte[] desktopNonce) {
        if (mobileNonce.length != Crypto.HANDSHAKE_NONCE_SIZE
                || desktopNonce.length != Crypto.HANDSHAKE_NONCE_SIZE) {
            throw new IllegalArgumentException(
                    "handshake nonces must be " + Crypto.HANDSHAKE_NONCE_SIZE + " bytes");
        }
        byte[] shared = Crypto.deriveSharedSecret(myPriv, peerPub);
        byte[] keyM2D = deriveKey(shared, mobileNonce, desktopNonce, Crypto.DIR_M2D);
        byte[] keyD2M = deriveKey(shared, mobileNonce, desktopNonce, Crypto.DIR_D2M);
        Arrays.fill(shared, (byte) 0);
        if (role == Role.MOBILE) {
            return new Session(keyM2D, keyD2M);
        }
        return new Session(keyD2M, keyM2D);
    }

    private static byte[] deriveKey(byte[] shared, byte[] mobileNonce, byte[] desktopNonce, String dir) {
        byte[] salt = new byte[mobileNonce.length + desktopNonce.length];
        System.arraycopy(mobileNonce, 0, salt, 0, mobileNonce.length);
        System.arraycopy(desktopNonce, 0, salt, mobileNonce.length, desktopNonce.length);
        byte[] info = (Crypto.INFO_PREFIX + dir).getBytes(StandardCharsets.UTF_8);
        return Hkdf.derive(shared, salt, info, Crypto.KEY_SIZE);
    }

    /**
     * Encrypt {@code plaintext} for the peer, binding {@code id}/{@code ts} as AEAD
     * associated data. Returns {@code nonce(24) || ciphertext}, ready to carry
     * verbatim as the envelope payload.
     */
    public byte[] seal(String id, long ts, byte[] plaintext) {
        return sealWith(Crypto.randomNonce(), id, ts, plaintext);
    }

    /** Seal with an explicit nonce — for deterministic interop vectors/tests. */
    byte[] sealWith(byte[] nonce, String id, long ts, byte[] plaintext) {
        byte[] ct = Crypto.aeadEncrypt(sendKey, nonce, Crypto.aad(id, ts), plaintext);
        byte[] wire = new byte[nonce.length + ct.length];
        System.arraycopy(nonce, 0, wire, 0, nonce.length);
        System.arraycopy(ct, 0, wire, nonce.length, ct.length);
        return wire;
    }

    /**
     * Decrypt a wire payload {@code nonce(24) || ciphertext} from the peer, checking
     * it against {@code id}/{@code ts}. Fails if the ciphertext, id, or ts was
     * tampered with.
     */
    public byte[] open(String id, long ts, byte[] wire) {
        if (wire.length < Crypto.NONCE_SIZE) {
            throw new IllegalArgumentException("ciphertext too short");
        }
        byte[] nonce = Arrays.copyOfRange(wire, 0, Crypto.NONCE_SIZE);
        byte[] ct = Arrays.copyOfRange(wire, Crypto.NONCE_SIZE, wire.length);
        byte[] pt = Crypto.aeadDecrypt(recvKey, nonce, Crypto.aad(id, ts), ct);
        // Reject duplicate delivery only after the AEAD has authenticated id/ts, so
        // forged metadata cannot advance the replay window (SG-02, FR-5.1).
        checkReplay(id, ts);
        return pt;
    }

    private synchronized void checkReplay(String id, long ts) {
        if (primed && ts < lastTs - replayWindowMillis) {
            throw new ReplayException("e2ee: timestamp outside replay window");
        }
        if (seen.containsKey(id)) {
            throw new ReplayException("e2ee: replay detected");
        }
        seen.put(id, ts);
        if (!primed || ts > lastTs) {
            lastTs = ts;
        }
        primed = true;
        long floor = lastTs - replayWindowMillis;
        for (Iterator<Map.Entry<String, Long>> it = seen.entrySet().iterator(); it.hasNext(); ) {
            if (it.next().getValue() < floor) {
                it.remove();
            }
        }
    }
}
